torch/linalg/__init__.py

```diff
@@ -140,6 +140,110 @@
     (tensor(3.7417), tensor(11.2250))
 """)
 
+eigh = _add_docstr(_linalg.linalg_eigh, r"""
+linalg.eigh(input, UPLO='L') -> tuple(Tensor, Tensor)
+
+This function returns eigenvalues and eigenvectors
+of a complex Hermitian (conjugate symmetric) or real symmetric matrix :attr:`input`
+represented by a namedtuple (eigenvalues, eigenvectors).
+
+This function calculates all eigenvalues (and vectors) of :attr:`input`
+such that :math:`\text{input} = V \text{diag}(e) V^H`, where :math:`e` is eigenvalues tensor
+and :math:`V` is eigenvectors tensor.
+
+Since the input matrix :attr:`input` is supposed to be Hermitian,
+only the lower triangular portion is used by default
+and the imaginary part of the diagonal will always be treated as zero.
+
+.. note:: The eigenvalues of real symmetric or complex Hermitian matrices are always real.
+
+.. note:: The eigenvalues/eigenvectors are computed using LAPACK/MAGMA routines ``_syevd``, ``_heevd``.
+
```

Comment:
This function causes a cross-device sync on CUDA, right? If so, this PR should add a warning here explaining that.

Reply:
That warning can be combined with a warning that this function will throw an error if a tensor does not meet the requirements.

Reply:
I added a note on cross-device synchronization for CUDA.

This function doesn't throw errors if the input is not Hermitian or symmetric because of

> Since :attr:`input` is assumed to be consisting of Hermitian matrices,
only the lower triangular portion of these matrices is used by default (``UPLO = 'L'``)  in computations
and the imaginary part of the diagonal will always be treated as zero.

[The documentation says](https://software.intel.com/content/www/us/en/develop/documentation/mkl-developer-reference-c/top/lapack-routines/lapack-least-squares-and-eigenvalue-problem-routines/lapack-least-squares-and-eigenvalue-problem-driver-routines/symmetric-eigenvalue-problems-lapack-driver-routines/syevd.html) the error code can be >0 indicating that the algorithm didn't converge. But I don't know what kind of input is needed to make it fail.

Reply:
Interesting... is our failure behavior consistent with NumPy's, then?

Reply:
Yes. NumPy raises `LinAlgError("Eigenvalues did not converge")` for `info > 0`.